### DISCUSSION TITLE
How to add more resources or update existing ones in runtime

### DISCUSSION THREAD

**Discussion by mrkolesnik** [r-simmer/simmer#302] [Q&A] [answered]
Hello

I have read almost all articles on the official web-site but couldn't find any example how to add more resources in runtime.

For example, I have defined some PC as resource (or IT server), which will process requests. Put some capacity to it, say capacity=2, implying that it has 2 cores for processing. But when a number of requests has been increased I would like more cores to it (i.e. increate capacity to 4) or add one more IT servers for processing in order to handle load.

**Comment by Enchufa2** [comment]
There are setters for all parameters. Here you are probably looking for [`set_capacity`](https://r-simmer.org/reference/set_capacity). And of course you have various [getters](https://r-simmer.org/reference/get_capacity) to check for the status of the resources.

There is one vignette showcasing this functionality: the [movie renege](https://r-simmer.org/articles/simmer-05-simpy#movie-renege) example.

**Comment by mrkolesnik** [comment]
Thanks @Enchufa2. It is quite useful and I can use it for vertical scaling in data center scenario. But I am also interested in `add_resource` in runtime, is that actually possible?

The scenario I'm currently building is based on next idea:

I can have heterogeneous environment, where can be different kind of processing servers (for example, less performable and more performable, like server with 1024 mb RAM and 1.33 GHz and other server with 2048 mb RAM and 2.6 GHz). In terms of simulation each resource can process requests with different speed.

But at first I need for simulation to have some amount of basic servers, then start simulation and when the load increases I would like to add advanced servers to the simulation or for example to shut them down in order to reduce maintenance cost.

**Reply by Enchufa2** [reply, marked as answer]
I never tried TBH. It _may_ work, but it wasn't intended to be called during a simulation. You can try, but why not adding all the resources that you may need beforehand, with capacity 0, and then increase their capacity or decrease it again to shut them down?

**Reply by mrkolesnik** [reply]
Thank you, @Enchufa2! It actually make sense. At least from the perspective of the simulation owner and from the point that resources are limited in the real world.
